File: leetcode/python/169_majority_element.py
import random


class Solution:
    def majorityElement(self, nums: list[int]) -> int:
        while True:
            i = random.randrange(len(nums))
            k = nums[i]
            count: int = 0
            for e in nums:
                if e == k:
                    count += 1
            if count > len(nums) // 2:
                return k

    def boyerMoore(self, nums: list[int]) -> int:
        count = 0
        candidate = None

        for num in nums:
            if count == 0:
                candidate = num
            count += (1 if num == candidate else -1)

        return candidate

s = Solution()
print(s.majorityElement([3,2,3]))
print(s.majorityElement([2,2,1,1,1,2,2]))

File: leetcode/rust/src/169_majority_element.rs
impl Solution {
    pub fn majority_element(nums: Vec<i32>) -> i32 {
        let mut count = 1;
        let mut candidate = nums[0];

        for n in nums[1..].iter() {
            if count == 0 {
                candidate = *n;
                count = 1;
            } else if *n == candidate {
                count += 1;
            } else {
                count -= 1;
            }
        }
        candidate
    }
}